Handshake interface

jannikluhn opened the pull request:
### What was wrong?

Identity scheme class only worked with ENR, but it's also responsible for the handshake.

### How was it fixed?

Implement the basic interface and add a placeholder so that we don't have to implement the cryptography as long as it isn't properly specified.

Also some minor stuff:
- rename node address to node id (both are used interchangeably in the spec, but address may collide with endpoint address)
- add method to access public key (we'll need that for the handshake)
- make node id and public key accessible on the ENR via properties

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://user-images.githubusercontent.com/29854669/61731364-9726dd00-ad7b-11e9-94ca-47d4e0a08610.jpg)

pipermerriam commented on p2p/discv5/identity_schemes.py:
```diff
@@ -65,51 +77,158 @@ class IdentityScheme(ABC):
+    @abstractmethod
+    def extract_public_key(cls, enr: "BaseENR") -> bytes:
+        """Retrieve the public key from an ENR."""
+        pass
```
We started moving over to using elipsis for abstract methods.

```python
@abstractmethod
def the_thing(self):
    ...
```

pipermerriam commented on p2p/discv5/identity_schemes.py:
```diff
@@ -65,51 +77,158 @@ class IdentityScheme(ABC):
-        public_key = PublicKey.from_compressed_bytes(enr[cls.public_key_enr_key])
-        return keccak(public_key.to_bytes())
+    def extract_public_key(cls, enr: "BaseENR") -> bytes:
+        return enr[cls.public_key_enr_key]
```
blind lookups like this always stand out to me as potential places for errors.  How do we know that this key is going to be present.  A quick glance at the `BaseENR` class didn't reveal anything obvious.

jannikluhn replied:
It's there, but somewhat hidden: `BaseENR.__init__` calls `IdentityScheme.validate_enr_structure` which (if `V4IdentityScheme` is used) checks that the key is present. I was trying to avoid the possibility of an exception here so that we can access the public key and node id via properties that behave nicely.

Do you think a `raise Exception("Unreachable: <comment>")` would be sufficient here to make this a bit more explicit or should I go back to methods instead of properties?

pipermerriam replied:
Probably fine but I'll point out that when things like this are not coupled (the validation and the access) it becomes easy to accidentally break in a refactor such as if the validation step were to move somewhere else.  I don't have a direct suggestion on how to restructure this so fine to leave as-is, or clean up later if you get some inspiration.

pipermerriam commented on p2p/discv5/typing.py:
```diff
@@ -1,7 +1,17 @@
+RandomBytesFn = Callable[[int], bytes]  # function that returns a number of random bytes
+
+
+class KeyAgreementResult(NamedTuple):
```
Is the term `KeyAgreement` from the spec?  If not what do you think about something like `SessionKeys`?

pipermerriam commented on tests/p2p/discv5/test_enr.py:
```diff
@@ -196,18 +205,25 @@ def test_signature_validation(mock_identity_scheme, identity_scheme_registry):
     assert mock_enr.identity_scheme is mock_identity_scheme
 
-    v4_enr = ENR(0, {b"id": b"v4"}, b"", identity_scheme_registry)
+    v4_enr = ENR(0, {b"id": b"v4", b"secp256k1": b"\x02" * 33}, b"", identity_scheme_registry)
```
I'd recommend going ahead and making yourself some factory classes for these objects under `p2p.tools.factories`

pipermerriam commented on tests/p2p/discv5/test_identity_schemes.py:
```diff
@@ -24,6 +24,16 @@
 
+@pytest.fixture
+def get_random_bytes():
+    random = Random(0)
```
probably better to go ahead and install [`pytest-randomly`](https://github.com/pytest-dev/pytest-randomly) and let it set the random seed.

pipermerriam commented on tests/p2p/discv5/test_identity_schemes.py:
```diff
@@ -81,5 +105,27 @@ def test_enr_node_address():
+    (b"\x02" * 34, False),
+))
+def test_handshake_public_key_validation(public_key, valid):
+    if valid:
```
I think we've leaned towards splitting these into two runs, one for the valid ones and one separate test fort he invalid ones.

jannikluhn commented:
Sorry for the delay. Changes since last review:

- Ellipsis instead of "pass" in abstract method bodies
- Renamed "KeyAgreementResult" to "SessionKeys" and "perform_key_agreement" to "compute_session_keys"
- In `V4IdentityScheme.extract_public_key(enr)`, I don't assume anymore that the ENR has been validated before (a proper exception is raised). Instead, this assumption is only made in `ENR.public_key`. It's still not ideal, but better than before, as now this issue is only present in one, not two classes. Also, I've added an `Exception("Invariant: ...")` in the property to document this slightly awkward coupling.
- Added id nonce signing and signature verification methods to the identity scheme which are also required for the handshake.
- I got rid of the `get_random_bytes` interface. The only reason why I thought it would be useful is for testing (so that we can replace the non-deterministic `os.urandom` we should use in production with a pseudo random alternative). But now I don't like it anymore as it increases the complexity of the interface for no good reason. Instead, I just mock out `os.urandom` in tests via a fixture.

What I have not done (yet):

- Use factories for creating ENRs. The problem here is that for most of the tests we actually test stuff in the ENRs `__init__` function, so what I'd actually have to do is use some kind of `ENRKwargFactory` which doesn't feel very useful for future tests. So since the existing tests already work fine I'd rather not spend more time on it at the moment to get it right, but revisit this for the next time I need an ENR for tests.
- Cleaned up the commit history. Will do so before I merge

Other than that, if I haven't missed anything, this is ready from my side.
